Networking: Update Mirror to v49.0.1

Ryan089 opened the pull request:
## Summary

Mirror has been updated to the most recent version (v49.0.1)

## Changes to Files

- Project settings was updated to reflect .Net 4.x compatibility level
- Multiple obsolete methods were replaced with the current methods
- Multiple prefabs had NetworkIdentities and NetworkTransforms removed from child objects.
- LoginNetworkManager was amended to fix error where the server would send a CharacterSelectMessage to the client.
- LoginNetworkManager was amended to more clearly define server and client responsibilities.
- SubstanceDisplay: Call to RpcUpdateDisplay was commented out due to exception in message handler (see Technical Notes)

## Technical Notes

While the new version of Mirror provides greater functionality and performance, it has some very particular features that we should be aware of:

- Any exceptions in a message handler (including Commands, Rpcs, Spawns and custom NetworkMessage handlers) will **cause the client to immediately disconnect**. This is to protect against malicious attackers. Nonetheless, it can be a pain to work with so we need to be careful to make any handlers robust.

- Mirror will no longer tolerate subordinate NetworkIdentities / NetworkTransforms on non-root prefabs. I have removed them for all of the items currently in the scene; however, it is likely that some prefabs exist with this issue.

- When loading this for the first time, you **may** have to delete your "obj" and "Temp" folders if it comes up with multiple obscure errors.

Ryan089 committed:
Direct replacement of Mirror folder with v49.0.1. Not yet functional.
(Test and Examples folders were excluded)

Ryan089 committed:
Updates ignoreAuthority to requiresAuthority.

Ryan089 committed:
Updated MessageBase to NetworkMessage

Ryan089 committed:
Updated ClientScene to NetworkClient

Ryan089 committed:
Updated various obsolete methods to ReadUInt and WriteUInt.

Ryan089 committed:
Removed calls to obsolete Logger.

Ryan089 committed:
NetworkMessages converted to structs

Ryan089 committed:
Adds the currentSender accessor to Mirror's RemoteCallHelper.

Ryan089 committed:
Added connectionFailed Action back onto Telepathy/Client.cs

Ryan089 committed:
Changed API compatibility level to .Net 4.x

Ryan089 committed:
Removed null-coalescing operator from a NetworkMessage instance
NetworkMessages are now structs rather than classes, so cannot be null.

Ryan089 committed:
NetworkConnection argument removed from NetworkClient.RegisterHandler() call.

Ryan089 committed:
Ensure NetworkClient is ready before calling AddPlayer.

Ryan089 committed:
Update Project Settings

Ryan089 committed:
Removed subordinate NetworkIdentities and NetworkTransforms from prefab children.
Also removed 'Missing' script from TankStorageUnit.prefab.

Ryan089 committed:
Commented out SubstanceDisplay.RpcUpdateDisplay()

This method suffered from a NullReferenceException in the message handler. The latest version of Mirror will automatically disconnect clients who have exceptions in NetworkMessages.

Ryan089 committed:
Updated LoginNetworkManager to more clearly define Client and Server methods.

SpawnPlayerAfterRoundStart() is a server-only method. Previously, it called SpawnPlayerWithoutLoginServer(), which should be a client-only method because it sends a CharacterSelectMessage. CharacterSelectMessages should only be sent from client to server. Because it was being sent from server to client instead, it was causing errors as there was not a registered handler of these messages on the client.

Ryan089 committed:
Removed a missing script from Human.

cosmiccoincidence commented:
I'm getting a horror monster effect when viewing the client player as the host. I explained with a video in discord.
I need someone to verify if this is in the PR or locally or what...

cosmiccoincidence commented:
Also, merge in master again for me now that the container rework is merged. It seems you have an outdated/conflicting file (LoginNetworkManager.cs)

Ryan089 committed:
Merge branch 'master' of https://github.com/RE-SS3D/SS3D into UpdateMirrorToV49.0.1

Conflicts:
  Assets/Engine/Server/Mirror/LoginNetworkManager.cs

Ryan089 committed:
Removed NetworkIdentities from Human prefab's hands.

Ryan089 committed:
Removed subordinate NetworkIdentities / NetworkTransforms:
- Vent.prefab
- SodaVendor.prefab
- Fridge.prefab

Ryan089 committed:
Removed subordinate NetworkIdentities from CellCharger.prefab.

Ryan089 committed:
Essential scene changes to MiniStation.unity.

Reloaded MiniStation tilemap from save. This is to remove any tile objects that were inadvertently added outside the Tilemap Editor.

Removed Janicart. The Janicart prefab is broken because of subordinate NetworkIdentities, and will be addressed in future work.

Ryan089 committed:
Put Burger.prefab back into the correct (item) layer.

Ryan089 committed:
Added NetworkStartPositions to MiniStation.

A somewhat disturbing bug had emerged where the Human prefab looked monstrous on the server (though looked normal on the client). It appeared that this was caused by the players instantiating in exactly the same position, as it did not occur when the players had moved away prior to the next one embarking. The four NetworkStartPositions added ensures that players will not routinely spawn as monsters.

It is possible that players *might* still look monstrous if they move onto another spawn point as a new player spawns there, but I think that is unlikely enough not to worry about for the moment.

cosmiccoincidence reviewed (approved):
Seems monster-free now.
